src/webgpu/shader/validation/parse/shadow_builtins.spec.ts

```diff
@@ -0,0 +1,1002 @@
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase({ requiredFeatures: ['shader-f16'] });
+  })
+  .fn(t => {
+    const data = kFloat16Tests[t.params.builtin];
+    const value = `let ${data.keyword} = 4;`;
+
+    const module_scope = t.params.inject === 'module' ? `var<private> ${data.keyword} : f16;` : ``;
+    const sibling_func = t.params.inject === 'sibling' ? value : ``;
+    const func = t.params.inject === 'function' ? value : ``;
+
+    const code = `
+enable f16;
+
+${module_scope}
+
+fn sibling() {
+  ${sibling_func}
+}
+
+@vertex
+fn vtx() -> @builtin(position) vec4f {
+  ${func}
+  ${data.src}
+  return vec4f(1);
+}
+    `;
+    const pass = t.params.inject === 'none' || t.params.inject === 'sibling';
+    t.expectCompileResult(pass, code);
+  });
+
+const kTextureTypeTests = {
+  texture_1d: {
+    keyword: `texture_1d`,
+    src: `var t: texture_1d<f32>;`,
+  },
+  texture_2d: {
+    keyword: `texture_2d`,
+    src: `var t: texture_2d<f32>;`,
+  },
+  texture_2d_array: {
+    keyword: `texture_2d_array`,
+    src: `var t: texture_2d_array<f32>;`,
+  },
+  texture_3d: {
+    keyword: `texture_3d`,
+    src: `var t: texture_3d<f32>;`,
+  },
+  texture_cube: {
+    keyword: `texture_cube`,
+    src: `var t: texture_cube<f32>;`,
+  },
+  texture_cube_array: {
+    keyword: `texture_cube_array`,
+    src: `var t: texture_cube_array<f32>;`,
+  },
+  texture_multisampled_2d: {
+    keyword: `texture_multisampled_2d`,
+    src: `var t: texture_multisampled_2d<f32>;`,
+  },
+  texture_depth_multisampled_2d: {
+    keyword: `texture_depth_multisampled_2d`,
+    src: `var t: texture_depth_multisampled_2d;`,
+  },
+  texture_external: {
+    keyword: `texture_external`,
+    src: `var t: texture_external;`,
+  },
+  texture_storage_1d: {
+    keyword: `texture_storage_1d`,
+    src: `var t: texture_storage_1d<rgba8unorm, read_write>;`,
+  },
+  texture_storage_2d: {
+    keyword: `texture_storage_2d`,
+    src: `var t: texture_storage_2d<rgba8unorm, read_write>;`,
+  },
+  texture_storage_2d_array: {
+    keyword: `texture_storage_2d_array`,
+    src: `var t: texture_storage_2d_array<rgba8unorm, read_write>;`,
+  },
+  texture_storage_3d: {
+    keyword: `texture_storage_3d`,
+    src: `var t: texture_storage_3d<rgba8unorm, read_write>;`,
+  },
+  texture_depth_2d: {
+    keyword: `texture_depth_2d`,
+    src: `var t: texture_depth_2d;`,
+  },
+  texture_depth_2d_array: {
+    keyword: `texture_depth_2d_array`,
+    src: `var t: texture_depth_2d_array;`,
+  },
+  texture_depth_cube: {
+    keyword: `texture_depth_cube`,
+    src: `var t: texture_depth_cube;`,
+  },
+  texture_depth_cube_array: {
+    keyword: `texture_depth_cube_array`,
+    src: `var t: texture_depth_cube_array;`,
+  },
+  sampler: {
+    keyword: `sampler`,
+    src: `var s: sampler;`,
+  },
+  sampler_comparison: {
+    keyword: `sampler_comparison`,
+    src: `var s: sampler_comparison;`,
+  },
+};
+
+g.test('shadow_hides_builtin_texture_type')
```

Comment:
nit: sampler and sampler_comparison aren't textures. `shadow_hides_handle_type`?

Reply:
done